### ISSUE TITLE
Section_2.3_3.2_Figure1.py

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/rarp-psa-trajectory-reproducibility#4]
"""Figure 1. Postoperative PSA trajectory cluster selection and characterization.

Manuscript alignment
--------------------
Methods: Section 2.3, PSA Trajectory Clustering
Results: Section 3.2, PSA Trajectory Clustering Results

Run 00_Common_Setup.py first.  This script rebuilds the eligible trajectory
cohort, evaluates K=2-6, fits the prespecified K=3 reference partition, and
creates all three panels and the final publication figure.
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from PIL import Image, ImageDraw, ImageFont, ImageOps
from scipy.ndimage import gaussian_filter1d
from sklearn.cluster import KMeans
from sklearn.impute import SimpleImputer
from sklearn.metrics import (
    adjusted_rand_score,
    calinski_harabasz_score,
    davies_bouldin_score,
    normalized_mutual_info_score,
    silhouette_score,
)

required_globals = (
    "ANALYSIS_READY_FILE",
    "OUTPUT_DIR",
    "ANALYSIS_READY_N",
    "TRAJECTORY_COHORT_N",
    "EXPECTED_CLUSTER_COUNTS",
    "EXPECTED_ANALYSIS_SHA256",
    "REFERENCE_SIGMA",
    "RANDOM_STATE",
    "KMEANS_N_INIT",
    "read_csv_robust",
    "verify_rows",
    "verify_sha256",
    "add_analysis_row_id",
    "psa_columns_in_time_order",
)
missing_globals = [name for name in required_globals if name not in globals()]
if missing_globals:
    raise RuntimeError(
        "Run 00_Common_Setup.py before this script. Missing: "
        + ", ".join(missing_globals)
    )

OUTDIR = OUTPUT_DIR / "Figure1"
OUTDIR.mkdir(parents=True, exist_ok=True)

K_GRID = (2, 3, 4, 5, 6)
STABILITY_SEEDS = tuple(range(50))

PANEL_A = OUTDIR / "Figure1A_Silhouette_by_K.png"
PANEL_B = OUTDIR / "Figure1B_Inertia_by_K.png"
PANEL_C = OUTDIR / "Figure1C_Mean_PSA_Trajectories.png"
FINAL_FIGURE = OUTDIR / "Figure1_PSA_Trajectory_Clustering.png"

# -----------------------------------------------------------------------------
# Cohort and frozen clustering features
# -----------------------------------------------------------------------------
verify_sha256(ANALYSIS_READY_FILE, EXPECTED_ANALYSIS_SHA256)
source = add_analysis_row_id(read_csv_robust(ANALYSIS_READY_FILE))
verify_rows(source, ANALYSIS_READY_N, "Analysis-ready dataset")

psa_columns = psa_columns_in_time_order(source)
psa_numeric = source[psa_columns].apply(pd.to_numeric, errors="coerce")
positive_count = ((psa_numeric > 0) & np.isfinite(psa_numeric)).sum(axis=1)
eligible = source.loc[positive_count >= 2].copy()
verify_rows(eligible, TRAJECTORY_COHORT_N, "Eligible trajectory cohort")

def clustering_features(frame: pd.DataFrame, sigma: float = REFERENCE_SIGMA) -> pd.DataFrame:
    psa = frame[psa_columns].apply(pd.to_numeric, errors="coerce")
    with np.errstate(divide="ignore", invalid="ignore"):
        log_psa = np.log(psa.replace(0, np.nan))

    # The reference clustering uses ordered adjacent observation columns.
    change = log_psa.diff(axis=1) / 1.0
    smoothed_array = np.vstack(
        [
            gaussian_filter1d(row.to_numpy(float), sigma=sigma, mode="nearest")
            for _, row in change.iterrows()
        ]
    )
    smoothed_change = pd.DataFrame(
        smoothed_array, index=change.index, columns=change.columns
    )

    features = pd.DataFrame(index=frame.index)
    features["max_change"] = smoothed_change.max(axis=1, skipna=True)
    features["mean_change"] = smoothed_change.mean(axis=1, skipna=True)
    features["first_positive_column_position"] = smoothed_change.apply(
        lambda row: int(np.argmax((row > 0).to_numpy(bool)))
        if np.any((row > 0).to_numpy(bool))
        else -1,
        axis=1,
    )
    return features

features = clustering_features(eligible)
imputer = SimpleImputer(strategy="median")
X = imputer.fit_transform(features)
if X.shape != (TRAJECTORY_COHORT_N, 3) or not np.isfinite(X).all():
    raise AssertionError(f"Unexpected clustering matrix: shape={X.shape}")

def fit_k(k: int, seed: int = RANDOM_STATE) -> tuple[KMeans, np.ndarray]:
    model = KMeans(
        n_clusters=k,
        n_init=KMEANS_N_INIT,
        random_state=seed,
    )
    return model, model.fit_predict(X)

# -----------------------------------------------------------------------------
# Panels A and B: cluster-number metrics
# -----------------------------------------------------------------------------
metric_rows = []
labels_by_k = {}
for k in K_GRID:
    model, labels = fit_k(k)
    labels_by_k[k] = labels
    counts = np.bincount(labels, minlength=k)
    metric_rows.append(
        {
            "K": k,
            "silhouette": float(silhouette_score(X, labels)),
            "calinski_harabasz": float(calinski_harabasz_score(X, labels)),
            "davies_bouldin": float(davies_bouldin_score(X, labels)),
            "inertia": float(model.inertia_),
            "cluster_sizes": "|".join(map(str, np.sort(counts))),
        }
    )

k_metrics = pd.DataFrame(metric_rows)
k_metrics["inertia_reduction_from_previous_pct"] = -100 * k_metrics["inertia"].pct_change()
k_metrics.to_csv(OUTDIR / "Figure1_K2_to_K6_metrics.csv", index=False)

fig, ax = plt.subplots(figsize=(7, 4.5))
ax.plot(k_metrics["K"], k_metrics["silhouette"], marker="o", linewidth=2)
ax.set(xlabel="K", ylabel="Silhouette score", title="Silhouette score across K")
ax.set_xticks(K_GRID)
fig.tight_layout()
fig.savefig(PANEL_A, bbox_inches="tight", facecolor="white")
plt.close(fig)

fig, ax = plt.subplots(figsize=(7, 4.5))
ax.plot(k_metrics["K"], k_metrics["inertia"], marker="o", linewidth=2)
ax.set(xlabel="K", ylabel="K-means inertia", title="K-means inertia across K")
ax.set_xticks(K_GRID)
fig.tight_layout()
fig.savefig(PANEL_B, bbox_inches="tight", facecolor="white")
plt.close(fig)

# -----------------------------------------------------------------------------
# Primary K=3 reference partition and Panel C
# -----------------------------------------------------------------------------
raw_k3 = labels_by_k[3]
raw_sizes = pd.Series(raw_k3).value_counts().to_dict()
if sorted(raw_sizes.values()) != sorted(EXPECTED_CLUSTER_COUNTS.values()):
    raise AssertionError(f"Unexpected K=3 cluster sizes: {raw_sizes}")

canonical_by_size = {78: 0, 984: 1, 372: 2}
raw_to_canonical = {
    int(raw_label): canonical_by_size[int(size)] for raw_label, size in raw_sizes.items()
}
canonical_labels = pd.Series(raw_k3).map(raw_to_canonical).to_numpy(int)
canonical_counts = pd.Series(canonical_labels).value_counts().sort_index().to_dict()
if canonical_counts != EXPECTED_CLUSTER_COUNTS:
    raise AssertionError(
        f"Expected canonical counts {EXPECTED_CLUSTER_COUNTS}; found {canonical_counts}"
    )

assignments = pd.DataFrame(
    {
        "Analysis_Row_ID": eligible["Analysis_Row_ID"].to_numpy(int),
        "Cluster": canonical_labels,
    }
)
assignments.to_csv(OUTDIR / "Figure1_Cluster_Assignments_1434.csv", index=False)

early_months = (1, 3, 6, 9, 12, 15, 18)
early_columns = [f"PSA POM{month}" for month in early_months]
trajectory_data = assignments.merge(
    eligible[["Analysis_Row_ID", *early_columns]],
    on="Analysis_Row_ID",
    how="left",
    validate="one_to_one",
)
trajectory_data[early_columns] = trajectory_data[early_columns].apply(
    pd.to_numeric, errors="coerce"
)
cluster_means = trajectory_data.groupby("Cluster")[early_columns].mean()
cluster_means.to_csv(OUTDIR / "Figure1C_Cluster_Mean_PSA.csv")

colors = {0: "#1f77b4", 1: "#ff7f0e", 2: "#2ca02c"}
fig, ax = plt.subplots(figsize=(10, 6))
for cluster in (0, 1, 2):
    ax.plot(
        early_months,
        cluster_means.loc[cluster].to_numpy(float),
        marker="o",
        linewidth=2.2,
        markersize=6,
        color=colors[cluster],
        label=f"Cluster {cluster} (n={EXPECTED_CLUSTER_COUNTS[cluster]})",
    )
ax.set_yscale("log")
ax.set_xticks(early_months)
ax.set_xlabel("Months after surgery")
ax.set_ylabel("Mean PSA (ng/mL)")
ax.set_title("Mean postoperative PSA trajectories")
ax.grid(True, which="major", alpha=0.45)
ax.legend(loc="upper left", frameon=True)
fig.tight_layout()
fig.savefig(PANEL_C, bbox_inches="tight", facecolor="white")
plt.close(fig)

# -----------------------------------------------------------------------------
# Random-initialization stability reported in Supplementary Table S2
# -----------------------------------------------------------------------------
stability_rows = []
for k in K_GRID:
    reference = labels_by_k[k]
    for seed in STABILITY_SEEDS:
        _, labels = fit_k(k, seed)
        stability_rows.append(
            {
                "K": k,
                "seed": seed,
                "ARI_vs_seed42": adjusted_rand_score(reference, labels),
                "NMI_vs_seed42": normalized_mutual_info_score(reference, labels),
                "silhouette": silhouette_score(X, labels),
            }
        )
pd.DataFrame(stability_rows).to_csv(
    OUTDIR / "Figure1_Random_Initialization_Stability.csv", index=False
)

# -----------------------------------------------------------------------------
# Final three-panel assembly
# -----------------------------------------------------------------------------
def load_rgb(path: Path) -> Image.Image:
    with Image.open(path) as image:
        return image.convert("RGB").copy()

def fit_inside(image: Image.Image, width: int, height: int) -> Image.Image:
    return ImageOps.contain(image, (width, height), method=Image.Resampling.LANCZOS)

panel_a = load_rgb(PANEL_A)
panel_b = load_rgb(PANEL_B)
panel_c = load_rgb(PANEL_C)

canvas_width = 4200
margin, column_gap, row_gap = 120, 100, 120
top_box_width = (canvas_width - 2 * margin - column_gap) // 2
panel_a = fit_inside(panel_a, top_box_width, 1300)
panel_b = fit_inside(panel_b, top_box_width, 1300)
panel_c = fit_inside(panel_c, 3150, 1750)

top_height = max(panel_a.height, panel_b.height)
canvas_height = margin + top_height + row_gap + panel_c.height + margin
canvas = Image.new("RGB", (canvas_width, canvas_height), "white")

a_x = margin + (top_box_width - panel_a.width) // 2
b_box_x = margin + top_box_width + column_gap
b_x = b_box_x + (top_box_width - panel_b.width) // 2
c_x = (canvas_width - panel_c.width) // 2
c_y = margin + top_height + row_gap

canvas.paste(panel_a, (a_x, margin))
canvas.paste(panel_b, (b_x, margin))
canvas.paste(panel_c, (c_x, c_y))

draw = ImageDraw.Draw(canvas)
font_path = Path("/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf")
font = ImageFont.truetype(str(font_path), 96) if font_path.exists() else ImageFont.load_default()
draw.text((30, 25), "A", fill="black", font=font)
draw.text((b_box_x - 90, 25), "B", fill="black", font=font)
draw.text((c_x - 105, c_y - 15), "C", fill="black", font=font)
canvas.save(FINAL_FIGURE, format="PNG", dpi=(300, 300), optimize=True)

# Published-value completion gate.
row2 = k_metrics.loc[k_metrics["K"] == 2].iloc[0]
row3 = k_metrics.loc[k_metrics["K"] == 3].iloc[0]
checks = {
    "K2 silhouette": (row2["silhouette"], 0.936341, 5e-6),
    "K3 silhouette": (row3["silhouette"], 0.913990, 5e-6),
    "K2-to-K3 inertia reduction": (
        row3["inertia_reduction_from_previous_pct"],
        68.043598,
        5e-5,
    ),
}
for label, (observed, expected, tolerance) in checks.items():
    if not np.isclose(observed, expected, atol=tolerance, rtol=0):
        raise AssertionError(f"{label}: expected {expected}; observed {observed}")

with Image.open(FINAL_FIGURE) as image:
    image.verify()

print(k_metrics.to_string(index=False))
print("\nCanonical cluster counts:", canonical_counts)
print("FIGURE 1 COMPLETION GATE PASSED")
print("Final figure:", FINAL_FIGURE)
